internal/ingress/annotations/backendprotocol/main.go

```diff
@@ -0,0 +1,62 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package backendprotocol
+
+import (
+	"regexp"
+	"strings"
+
+	"github.com/golang/glog"
+	extensions "k8s.io/api/extensions/v1beta1"
+
+	"k8s.io/ingress-nginx/internal/ingress/annotations/parser"
+	"k8s.io/ingress-nginx/internal/ingress/resolver"
+)
+
+var (
+	validProtocols = regexp.MustCompile(`^(HTTP|HTTPS|AJP|GRPC|GRPCS)$`)
+)
+
+type backendProtocol struct {
+	r resolver.Resolver
+}
+
+// NewParser creates a new backend protocol annotation parser
+func NewParser(r resolver.Resolver) parser.IngressAnnotation {
+	return backendProtocol{r}
+}
+
+// ParseAnnotations parses the annotations contained in the ingress
+// rule used to indicate the backend protocol.
+func (a backendProtocol) Parse(ing *extensions.Ingress) (interface{}, error) {
+	if ing.GetAnnotations() == nil {
+		return "HTTP", nil
+	}
+
+	proto, err := parser.GetStringAnnotation("backend-protocol", ing)
+	if err != nil {
```

Comment:
should we log the error here?

Reply:
No? (we are not doing that in any other annotation that requires to return a default value)

Reply:
I did not mean return error, I meant log a warning just like we do at https://github.com/kubernetes/ingress-nginx/pull/2871/files#diff-42abcd9476cd9422cd55e93f503bd2e2R57

Reply:
That would log an error every time this annotation is unset. Expect a lot of verbosity :)

Reply:
oh strange that it returns error when the annotation is unset... kcco then

Reply:
I had the same question the first time I went through that code: https://github.com/kubernetes/ingress-nginx/blob/cdbb9bdd76e1d7f61bca6dac6ff6624b6834ed9e/internal/ingress/annotations/parser/main.go#L52-L58
I guess @aledbf has an explanation :)